File: contains-duplicate/contains-duplicate.cpp
#include <stdio.h>
#include <assert.h>
#include <iostream>
#include <vector>
#include <unordered_set>

using namespace std;

class Solution {
public:
    bool containsDuplicate(vector<int>& nums) {
        unordered_set<int> s;
        for (int i : nums) {
            if (s.count(i) > 0) {
                return true;
            }
            s.insert(i);
        }
        return false;
    }
};


int main(void) {
    Solution sol = Solution();
    vector<int> t1{ 1,2,3,1 };
    assert(sol.containsDuplicate(t1));

    vector<int> t2{ 1,2,3,4 };
    assert(!sol.containsDuplicate(t2));

    return 0;
}

File: n150/contains-duplicate/contains-duplicate.rs
// https://leetcode.com/problems/contains-duplicate
struct Solution;

use std::collections::HashSet;

impl Solution {
    pub fn contains_duplicate(nums: Vec<i32>) -> bool {
        let mut set = HashSet::with_capacity(nums.len());
        for num in nums.into_iter() {
            if set.contains(&num) {
                return true;
            }
            set.insert(num);
        }

        false
    }
}

fn main() {
    let tests = vec![
        (
            vec![1,2,3,1],
            true,
        ),
        (
            vec![1,2,3,4],
            false
        ),
        (
            vec![1,1,1,3,3,4,3,2,4,2],
            true,
        )
    ];

    for (nums, solution) in tests {
        let result = Solution::contains_duplicate(nums);
        assert_eq!(result, solution);
    }

}
